### ISSUE TITLE
Example program

### ISSUE THREAD

**Issue by likeafox** [dgrunwald/rust-cpython#64]
I'm running into trouble trying to get the first example program on this project's main page to work for me.  It builds OK, but running it does nothing but crash.  I don't even know where to begin troubleshooting this:

```
cargo run
Running `target\debug\yay_python.exe`
error: Process didn't exit successfully: `target\debug\yay_python.exe` (exit code: 3221225477)
```

I think it is failing at  `let sys = py.import("sys").unwrap();`  because the result is identical when everything after that line is commented out.

My environment:
Windows
Rust 1.12.1 gnu 32 bit
Python 3.4.4 32 bit

What could I be doing wrong?

**Comment by dgrunwald** [comment]
rust-cpython currently only works on Linux and OSX. I don't know how to correctly link with the python DLL on Windows.
`windows-gnu` produces weirdly corrupt binaries (which might explain the crash you are seeing); and `windows-msvc` doesn't link at all (see #10).

**Comment by AraHaan** [comment]
Well to link it cant you just rename a copy of ``pythonxx.lib`` to ``pythonxx.dlib``? from wherever they have python installed at?

**Comment by AraHaan** [comment]
Another idea is to look for the python libs in ``%SystemDrive%\Pythonlibs`` and then have the end user copy the ``pythonxx.lib`` files to that folder.

**Comment by dgrunwald** [comment]
Finding the .lib is not the problem.

**Comment by AraHaan** [comment]
hmm
When what part on the linking that lib file is?

**Comment by dgrunwald** [comment]
See my previous comment in this thread.
Or in short: `windows-gnu` produces corrupt binaries; `windows-msvc` fails due to missing `__declspec(dllimport)`.

**Comment by AraHaan** [comment]
hmm maybe what you need is for it to differentiate from the msvc build of rust from the gnu one.

**Comment by AraHaan** [comment]
@dgrunwald I was told by a rust contributor on https://github.com/rust-lang/rust/issues/38285#issuecomment-266233079 that you can in the build script get if it is an ``msvc`` or a ``gnu`` build of rust. This might help fix the issue with importing ``pythonxx.dll``

**Comment by AraHaan** [comment]
ok, I think I made a somewhat better rewrite on the build script:

```rust
extern crate target_build_utils;
use std::env;
use std::io::Write;
use target_build_utils::TargetInfo;

const CFG_KEY: &'static str = "py_sys_config";

#[cfg(feature="python27-sys")]
const PYTHONSYS_ENV_VAR: &'static str = "DEP_PYTHON27_PYTHON_FLAGS";

#[cfg(feature="python3-sys")]
const PYTHONSYS_ENV_VAR: &'static str = "DEP_PYTHON3_PYTHON_FLAGS";

fn main() {
    // python{27,3.x}-sys/build.rs passes python interpreter compile flags via 
    // environment variable (using the 'links' mechanism in the cargo.toml).
    let flags = match env::var(PYTHONSYS_ENV_VAR) {
        Ok(flags) => flags,
        Err(_) => {
            writeln!(std::io::stderr(),
                "Environment variable {} not found - this is supposed to be \
                 exported from the pythonXX-sys dependency, so the build chain is broken",
                PYTHONSYS_ENV_VAR).unwrap();
            std::process::exit(1);
        }
    };
    let target = TargetInfo::new().expect("could not get target");
    if target.target_vendor() == "x86_64-pc-windows-msvc" {
        // MSVC Building here.
        //This is where __declspec(dllimport) can be defined.
    }
    if target.target_vendor() == "i686-pc-windows-msvc" {
        // MSVC Building here.
        //This is where __declspec(dllimport) can be defined.
    }
    if target.target_vendor() == "x86_64-pc-windows-gnu" {
        // gnu Building here.
    }
    if target.target_vendor() == "i686-pc-windows-gnu" {
        // gnu Building here.
    }

    if flags.len() > 0 {
        for f in flags.split(",") {
            // write out flags as --cfg so that the same #cfg blocks can be used
            // in rust-cpython as in the -sys libs
            let key_and_val: Vec<&str> = f.split("=").collect();
            let key = key_and_val[0];
            let val = key_and_val[1];
            if key.starts_with("FLAG") {
                println!("cargo:rustc-cfg={}=\"{}\"", CFG_KEY, &key[5..])
            } else {
                println!("cargo:rustc-cfg={}=\"{}_{}\"", CFG_KEY, &key[4..], val);
            }
        }
    }
}
```

With this it would mean also adding this to Cargo.toml:

```toml
[build-dependencies]
target_build_utils = "*"
```

**Comment by dgrunwald** [comment]
Linking on `windows-msvc` should be working now, assuming you are using the latest nightly -- the Rust issues with `dllimport` were just fixed in rustc 1.15.0-nightly (8f02c429a 2016-12-15)
